### ISSUE TITLE
CDS metadata keys should be filter instance names, not filter implementation names

### ISSUE THREAD

**Issue by markdroth** [envoyproxy/envoy#34251]
In #12274, we changed the keys in `typed_per_filter_config` to match the HTTP filter *instance* name (the [`HttpFilter.name` field](https://github.com/envoyproxy/envoy/blob/7436690884f70b5550b6953988d05818bae3d087/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L1149)) instead of the legacy filter implementation name (e.g., "envoy.filters.http.gcp_authn").  However, the CDS metadata keys are still using the legacy names ([example](https://github.com/envoyproxy/envoy/blob/7436690884f70b5550b6953988d05818bae3d087/source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc#L51)).

I think we should make the same change for CDS metadata keys that we did for `typed_per_filter_config`, so that it becomes possible to specify different metadata for multiple instances of the same filter.

**Comment by wbpcode** [comment]
I think not only the CDS metadata, we have no restriction to the using of any metadata. It's completely depend on the filter implementation. And it's much harder to change this point. 🤔

**Comment by markdroth** [comment]
You're right, this is not just an issue for CDS, but for metadata in any resource.  Updated the issue title accordingly.

I think we should probably build a common library for finding metadata entries applicable to a given filter in various parts of the config (e.g., routes, clusters, endpoints, etc), and then we should modify all existing filters to use that common library.

**Comment by markdroth** [comment]
For context, the reason I filed this issue is that I am proposing a feature in gRPC that will be the first to use CDS metadata (see [gRFC A83: xDS GCP Authentication Filter](https://github.com/grpc/proposal/pull/438)), and I want to make sure we agree on the right semantics before we implement it.  I don't think we want Envoy and gRPC to have incompatible interpretations of this data, but at the same time, Envoy's existing behavior seems broken to me, and I don't want to implement this broken behavior in gRPC and then have to fix it later.

**Comment by adisuissa** [comment]
I agree that it makes sense. We probably want Envoy to be backward compatible, i.e., search b a filter name, and if not found, search for a filter of that type (keeping the current behavior).

**Comment by markdroth** [comment]
Yup, absolutely.  We can use basically the same migration story that we did for #12274.

**Comment by markdroth** [comment]
When we do this, we need to make sure that we get the CDS validation semantics right.  Here's pseudo-code for the algorithm that we're going to implement in gRPC:

https://github.com/grpc/proposal/blob/0cae3c5edf8fc072c97b1f2e1bd67c9fe0ac7c8d/A83-xds-gcp-authn-filter.md#xds-cluster-metadata
